src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SslOverTdsStream.NetCoreApp.cs

```diff
@@ -0,0 +1,164 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class SslOverTdsStream
+    {
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadInternal(new Memory<byte>(buffer, offset, count), default, async: false).GetAwaiter().GetResult();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken token)
+            => ReadInternal(new Memory<byte>(buffer, offset, count), token, async: true).AsTask();
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+            => ReadInternal(buffer, cancellationToken, async: true);
+
+        public override void Write(byte[] buffer, int offset, int count)
+            => WriteInternal(new ReadOnlyMemory<byte>(buffer, offset, count), default, async: true).GetAwaiter().GetResult();
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken token)
+            => WriteInternal(new ReadOnlyMemory<byte>(buffer, offset, count), token, async: true).AsTask();
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            => WriteInternal(buffer, cancellationToken, async: true);
+
+        /// <summary>
+        /// Read Internal is called synchronosly when async is false
+        /// </summary>
+        private async ValueTask<int> ReadInternal(Memory<byte> buffer, CancellationToken token, bool async)
+        {
+            int readBytes = 0;
+            int count = buffer.Length;
+            byte[] packetData = new byte[count < TdsEnums.HEADER_LEN ? TdsEnums.HEADER_LEN : count];
+
+            if (_encapsulate)
+            {
+                if (_packetBytes == 0)
+                {
+                    // Account for split packets
+                    while (readBytes < TdsEnums.HEADER_LEN)
+                    {
+                        readBytes += async ?
```

Comment:
This can infinite loop if the connection is closed and the read returns 0 and doesn't throw, Are closed connections guaranteed to throw or return non-zero?

Reply:
Hmm... the existing loop also has this issue.

Will fix in both. Good spot

Reply:
Someone else mentioned it elsewhere before I just haven't got around to looking at it in any depth yet so since you're here anyway I thought it worth mentioning. 

There may not be a clear way to "fix" it from what I remember. I'm not sure what should happen if you end up in a partial packet read because it's below ssl stream which is going to try and decode anything you return in the buffer which can error that looks like a transport error when in fact it's connection closed. See what you think.